knowhere/common/FileManager.h

```diff
@@ -0,0 +1,64 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+#pragma once
+
+#include <optional>
+
+namespace knowhere {
+
+/**
+ * @brief This FileManager is used to manager file, including its replication, backup, ect. 
+ * It will act as a cloud-like client, and Knowhere need to call load/add to better support
+ * distribution of the whole service.
+ * 
+ * (TODO) we need support finer granularity file operator (read/write), 
+ * so Knowhere doesn't need to offer any help for service in the future .
+ */
+class FileManager {
+    /**
+     * @brief Load a file to the local disk, so we can use stl lib to operate it.
+     *
+     * @param filename
+     * @return false if any error, or return true.
+     */
+    virtual bool
+    load_file(const std::string& filename) noexcept;
+
+    /**
+     * @brief Add file to FileManager to manipulate it.
+     *
+     * @param filename
+     * @return false if any error, or return true.
+     */
+    virtual bool
+    add_file(const std::string& filename) noexcept;
+
+    /**
+     * @brief Check if a file exists.
+     *
+     * @param filename
+     * @return std::nullopt if any error, or return if the file exists.
+     */
+    virtual std::optional<bool>
```

Comment:
why use "std::optional<bool>"

Reply:
I am trying to avoid exceptions. It is a common way to use `std::optional` to represent simple error code, in which `nullopt` indicates error.